### PR TITLE
Fix up common ReaderId pattern

### PR DESCRIPTION
**Pull request by Xaeroxe**
I realized there's a better way to use `ReaderId`s in `System`s than we have been using, so I set out to implement this pattern across the engine.  Functionality should be identical, we just don't need `Option<ReaderId>` any more.

### CODE REVIEW

**Comment by udoprog** [comment]
A more general question: Is there ever a circumstance where we don't want to call `Self::SystemData::setup`?

If so, this seems like it could be done in shred dedicating the setup hook fully to custom code.
Making this kind of customization less noisy and prone to errors.

**Comment by Xaeroxe** [comment]
The only reason I could see to not do that is that then every `dispatch()` call has to check a boolean to see if `setup` has already been run or not.  Maybe a good note for @torkleyy to tuck in with Specs 2.0.

File: amethyst_core/src/transform/systems.rs [hunk 1]
```diff
@@ -84,34 +79,31 @@ impl<'a> System<'a> for TransformSystem {
...
-        )
-            .join()
         {
-            self.global_modified.add(entity.id());
-            global.0 = local.matrix();
-            debug_assert!(
-                global.is_finite(),
-                format!("Entity {:?} had a non-finite `Transform`", entity)
-            );
+            let (local_modified, global_modified) = data.borrow_parts();
```
[the remark this answers is not in the document]
Comment (Xaeroxe) [reply amethyst_core/src/transform/systems.rs]: Maybe, I'll give it a shot.

[the remark this answers is not in the document]
Comment (Xaeroxe) [reply amethyst_core/src/transform/systems.rs]: That worked, and is objectively superior.  Thanks!

**Comment by Xaeroxe** [comment]
It does, but we also don't have this problem of a partially initialized structure.  We also don't have to use as many unwraps and expects.

**Comment by Xaeroxe** [comment]
This is step 1 of a larger goal of mine, which is to get data out of systems.  One of the biggest reasons to use an ECS is for that separation of concerns, systems should not store state.  If we succeed in removing all state from all systems then I believe we'll have a simpler initialization story.

**Comment by torkleyy** [comment]
Interesting. This is part of a bigger issue; I've also thought about not having fields for systems at all and making everything a resource, but I'm unsure how many issues that actually solves or creates.

One of the issues that also comes up here is that as soon as you re-build your dispatcher (without this system), you'll get a major memory leak because the `EventChannel` keeps expanding its buffer due to how `ReaderId`s work. Some of that could be improved, but I think that's a general issue that comes from using resources from this. I see some advantages as well, but this means we can't merge this PR as it is.

**Comment by udoprog** [comment]
> One of the issues that also comes up here is that as soon as you re-build your dispatcher (without this system), you'll get a major memory leak because the EventChannel keeps expanding its buffer due to how ReaderIds work. Some of that could be improved, but I think that's a general issue that comes from using resources from this. I see some advantages as well, but this means we can't merge this PR as it is.

I'm not sure I understand the details of this memory leak: Wouldn't the old systems be dropped and the `ReaderId`'s released in this case?

**Comment by torkleyy** [comment]
Right now: yes
If the reader id is in a resource: No, because it will persist there

**Comment by Xaeroxe** [comment]
@torkleyy that's actually a really good point.  I'll give it some thought.

**Comment by Xaeroxe** [comment]
Since I don't have a good solution @torkleyy 's problem right now I'm going to close this PR.

**Pull request closed without merging**